src/models/vae.jl

```diff
@@ -0,0 +1,150 @@
+using Flux
+using ConditionalDists
+using GenerativeModels
+using ValueHistories
+using MLDataPattern: RandomBatches
+using Distributions
+using StatsBase
+using Random
+
+"""
+	safe_softplus(x::T)
+
+Safe version of softplus.	
+"""
+safe_softplus(x::T) where T  = softplus(x) + T(0.000001)
+
+
+"""
+	vae_constructor(;idim::Int=1, zdim::Int=1, activation = "relu", hdim=128, nlayers=3, kwargs...)
+"""
+function vae_constructor(;idim::Int=1, zdim::Int=1, activation = "relu", hdim=128, nlayers::Int=3, init_seed::Int=nothing, kwargs...)
+	(nlayers < 2) ? error("Less than 3 layers are not supported") : nothing
+	# function from string
+	act = eval(Meta.parse(activation))
+
+	# if seed is given, set it
+	(init_seed != nothing) ? Random.seed!(init_seed) : nothing
+	
+	# construct the model
+	# encoder - diagonal covariance
+	encoder_map = Chain(
+		Dense(idim,hdim,act),
+		[Dense(hdim, hdim, act) for _ in 1:nlayers-2]...,
+		ConditionalDists.SplitLayer(hdim, [zdim, zdim], [identity, safe_softplus])
+		)
+	encoder = ConditionalMvNormal(encoder_map)
+	# decoder - we will optimize only a shared scalar variance for all dimensions
+	decoder_map = Chain(
+		Dense(zdim,hdim,act),
+		Dense(hdim,hdim,act),
+		ConditionalDists.SplitLayer(hdim, [idim, 1], [identity, safe_softplus])
+		)
+	decoder = ConditionalMvNormal(decoder_map)
+
+	# reset seed
+	(init_seed != nothing) ? Random.seed!() : nothing
+
+	# get the vanilla VAE
+	model = VAE(zdim, encoder, decoder)
+end
+
+"""
+	loss(model::GenerativeModels.VAE, x[, batchsize])
+
+Negative ELBO for training of a VAE model.
+"""
+loss(model::GenerativeModels.VAE, x) = -elbo(model, x)
+# version of loss for large datasets where
+loss(model::GenerativeModels.VAE, x, batchsize::Int) = 
+	mean(map(y->loss(model,y), Flux.Data.DataLoader(x, batchsize=batchsize)))
+
+"""
+	StatsBase.fit!(model::GenerativeModels.VAE, data::Tuple; max_train_time=82800, lr=0.001, 
+		batchsize=64, patience=30, check_interval::Int=10, kwargs...)
+"""
+function StatsBase.fit!(model::GenerativeModels.VAE, data::Tuple; max_train_time=82800, lr=0.001, 
+	batchsize=64, patience=30, check_interval::Int=10, kwargs...)
+	history = MVHistory()
+	opt = ADAM(lr)
+
+	tr_model = deepcopy(model)
+	ps = Flux.params(tr_model)
+	_patience = patience
+
+	tr_x = data[1][1]
+	val_x = data[2][1][:,data[2][2] .== 0]
+	val_N = size(val_x,2)
+
+	# on large datasets, batching loss is faster
+	best_val_loss = Inf
+	i = 1
+	start_time = time() # end the training loop after 23hrs
+	for batch in RandomBatches(tr_x, batchsize)
+		batch_loss = 0f0
+		gs = gradient(() -> begin 
+			batch_loss = loss(tr_model,batch)
+		end, ps)
+	 	Flux.update!(opt, ps, gs)
+
+		# validation/early stopping
+		val_loss = (val_N > 5000) ? loss(tr_model, val_x, 256) : loss(tr_model, val_x)
+		
+		(i%check_interval == 0) ? (@info "$i - loss: $(batch_loss) (batch) | $(val_loss) (validation)") : nothing
```

Comment:
Is `check_interval` is used only for printing or did you intend to check also the validation loss also less often? I'm asking just if it makes sense to put it as a parameter, because currently it's just a debugging technicality.

Reply:
Yeah it's only used for printing so that there is less output. I can see it being an argument of the function for debugging of model, not really a parameter to be sampled from.